src/privatesend-client.cpp

```diff
@@ -1316,13 +1318,17 @@ bool CPrivateSendClient::CreateDenominated(const CompactTallyItem& tallyItem, bo
             nFeeRet, nChangePosRet, strFail, &coinControl, true, ONLY_NONDENOMINATED_NOT1000IFMN);
     if(!fSuccess) {
         LogPrintf("CPrivateSendClient::CreateDenominated -- Error: %s\n", strFail);
-        // TODO: return reservekeyDenom here
+        for(auto key : reservekeyDenomVec)
+            key->ReturnKey();
         reservekeyCollateral.ReturnKey();
+        LogPrintf("CPrivateSendClient::CreateDenominated -- %d keys returned\n", reservekeyDenomVec.size() + 1);
         return false;
     }
 
-    // TODO: keep reservekeyDenom here
+    for(auto key : reservekeyDenomVec)
+        key->KeepKey();
     reservekeyCollateral.KeepKey();
+    LogPrintf("CPrivateSendClient::CreateDenominated -- %d keys keeped\n", reservekeyDenomVec.size() + 1);
 
     if(!pwalletMain->CommitTransaction(wtx, reservekeyChange)) {
         LogPrintf("CPrivateSendClient::CreateDenominated -- CommitTransaction failed!\n");
```

Comment:
The thing that was failing here was always `CreateTransaction` in practice. `CommitTransaction` can only fail if it's not accepted to mempool - `must not fail. The transaction has already been signed and recorded.`. However, to make sure we still actually have inputs we are trying to spend, I guess we have to `LOCK2(cs_main, pwalletMain->cs_wallet);` at the start of `CPrivateSendClient::CreateDenominated()`. Other places with Create/Commit have the same issue I guess, but should be fixed in separate PR.